pkg/alertmanager/operator.go

```diff
@@ -183,13 +281,150 @@ func (c *Operator) addHandlers() {
 		DeleteFunc: c.handleAlertmanagerDelete,
 		UpdateFunc: c.handleAlertmanagerUpdate,
 	})
+	c.alrtCfgInfs.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleAlertmanagerConfigAdd,
+		DeleteFunc: c.handleAlertmanagerConfigDelete,
+		UpdateFunc: c.handleAlertmanagerConfigUpdate,
+	})
+	c.secrInfs.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.handleSecretAdd,
+		DeleteFunc: c.handleSecretDelete,
+		UpdateFunc: c.handleSecretUpdate,
+	})
 	c.ssetInfs.AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    c.handleStatefulSetAdd,
 		DeleteFunc: c.handleStatefulSetDelete,
 		UpdateFunc: c.handleStatefulSetUpdate,
 	})
 }
 
+func (c *Operator) handleAlertmanagerConfigAdd(obj interface{}) {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig added")
+		c.metrics.TriggerByCounter(monitoringv1beta1.AlertmanagerConfigKind, "add").Inc()
+
+		c.enqueueForMonitorNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleAlertmanagerConfigUpdate(old, cur interface{}) {
+	if old.(*monitoringv1beta1.AlertmanagerConfig).ResourceVersion == cur.(*monitoringv1beta1.AlertmanagerConfig).ResourceVersion {
+		return
+	}
+
+	o, ok := c.getObject(cur)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig updated")
+		c.metrics.TriggerByCounter(monitoringv1beta1.AlertmanagerConfigKind, "update").Inc()
+
+		c.enqueueForMonitorNamespace(o.GetNamespace())
+	}
+}
+
+func (c *Operator) handleAlertmanagerConfigDelete(obj interface{}) {
+	o, ok := c.getObject(obj)
+	if ok {
+		level.Debug(c.logger).Log("msg", "AlertmanagerConfig delete")
+		c.metrics.TriggerByCounter(monitoringv1beta1.AlertmanagerConfigKind, "delete").Inc()
+
+		c.enqueueForMonitorNamespace(o.GetNamespace())
+	}
+}
+
+// TODO: Do we need to enque secrets just for the namespace or in general?
```

Comment:
Is this TODO still valid before we merge the PR?

Reply:
I plagiarized this TODO when I plagiarised this function from here :see_no_evil: https://github.com/prometheus-operator/prometheus-operator/blob/9a0823d7d1c2a1034a3ac7a05118cc02c1d0df52/pkg/prometheus/operator.go#L891

I could be wrong, but here's how I see it:

* We should enqueue reconciliation for an `Alertmanager` if any of the `Secrets` referenced in its spec are deleted (it has a `secrets` field which is a list, `configSecret` field, `containers` or `initContainers` fields may reference a Secret to mount, etc). These all have to be in the same namespace as the Alertmanager. This function, enqueuing all Alertmanagers in the namespace, covers these cases, I believe.

* We _may_ want to also enqueue reconciliation for an `Alertmanager`, if there's an AlertmanagerConfig selected by it that references a Secret that's deleted (upon reconciliation, if the Secret is missing for an AlertmanagerConfig, then that AlertmanagerConfig will be excluded when building the config file). If that's what we want, then I think it would _either_ mean enqueuing all Alertmanagers in _all_ namespaces, for any Secret changes in any AlertmanagerConfig namespaces; or determining if an Alertmanager should be reconciled by checking if the Secret is referenced in either the Alertmanager spec, or in any AlertmanagerConfigs that it selects.

I'm starting to confuse myself. In any case, if we want the second option in addition to the first, I'm not quite sure how to achieve that here. :thinking: